contrib/go-redis/redis.v8/redis.go

```diff
@@ -0,0 +1,162 @@
+		tracer.Tag(ext.TargetHost, p.host),
+		tracer.Tag(ext.TargetPort, p.port),
+		tracer.Tag("out.db", p.db),
+		tracer.Tag("redis.raw_command", raw),
+		tracer.Tag("redis.args_length", strconv.Itoa(length)),
+	}
+	if !math.IsNaN(p.config.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, p.config.analyticsRate))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, "redis.command", opts...)
+	return ctx, nil
+}
+
+func (ddh *datadogHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	var span tracer.Span
+	span, _ = tracer.SpanFromContext(ctx)
+	var finishOpts []ddtrace.FinishOption
+	errRedis := cmd.Err()
+	if errRedis != redis.Nil {
+		finishOpts = append(finishOpts, tracer.WithError(errRedis))
+	}
+	span.Finish(finishOpts...)
+	return nil
+}
+
+func (ddh *datadogHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	raw := commandsToString(cmds)
+	parts := strings.Split(raw, " ")
+	length := len(parts) - 1
+	p := ddh.params
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeRedis),
+		tracer.ServiceName(p.config.serviceName),
+		tracer.ResourceName(parts[0]),
+		tracer.Tag(ext.TargetHost, p.host),
+		tracer.Tag(ext.TargetPort, p.port),
+		tracer.Tag("out.db", p.db),
+		tracer.Tag("redis.raw_command", raw),
+		tracer.Tag("redis.args_length", strconv.Itoa(length)),
+		tracer.Tag(ext.ResourceName, raw),
+		tracer.Tag("redis.pipeline_length", strconv.Itoa(len(cmds))),
+	}
+	if !math.IsNaN(p.config.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, p.config.analyticsRate))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, "redis.command", opts...)
+	return ctx, nil
+}
+
+func (ddh *datadogHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	var span tracer.Span
+	span, _ = tracer.SpanFromContext(ctx)
+	var finishOpts []ddtrace.FinishOption
+	for _, cmd := range cmds {
+		errCmd := cmd.Err()
+		if errCmd != redis.Nil {
+			finishOpts = append(finishOpts, tracer.WithError(errCmd))
+		}
+	}
+	span.Finish(finishOpts...)
+	return nil
+}
+
+// commandsToString returns a string representation of a slice of redis Commands, separated by newlines.
+func commandsToString(cmds []redis.Cmder) string {
+	var b bytes.Buffer
+	for _, cmd := range cmds {
+		b.WriteString(cmderToString(cmd))
+		b.WriteString("\n")
+	}
+	return b.String()
+}
+
+func cmderToString(cmd redis.Cmder) string {
+	// We want to support multiple versions of the go-redis library. In
+	// older versions Cmder implements the Stringer interface, while in
+	// newer versions that was removed, and this String method which
+	// sometimes returns an error is used instead. By doing a type assertion
+	// we can support both versions.
+	if stringer, ok := cmd.(fmt.Stringer); ok {
+		return stringer.String()
+	}
+
+	args := cmd.Args()
+	if len(args) == 0 {
+		return ""
+	}
+	if str, ok := args[0].(string); ok {
+		return str
+	}
+	return ""
+}
```

Comment:
Cmder implements fmt.Stringer at v8. https://pkg.go.dev/github.com/go-redis/redis/v8#Cmder
We don't need this hack.